src/coreclr/jit/helperexpansion.cpp

```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+    assert(call->gtArgs.CountUserArgs() == 5);
+
+    // First, list all arguments of the intrinsic call, the signature is:
+    //
+    //   int GetUtf8Bytes(ref char srcPtr, int srcLen, ref byte dstPtr, int dstLen)
+    //
+    GenTree* thisArg = call->gtArgs.GetUserArgByIndex(0)->GetNode();
+    GenTree* srcPtr  = call->gtArgs.GetUserArgByIndex(1)->GetNode();
+    GenTree* srcLen  = call->gtArgs.GetUserArgByIndex(2)->GetNode();
+    GenTree* dstPtr  = call->gtArgs.GetUserArgByIndex(3)->GetNode();
+    GenTree* dstLen  = call->gtArgs.GetUserArgByIndex(4)->GetNode();
+
+    // We're interested in a case when srcPtr is a string literal and srcLen is a constant
+
+    ssize_t               strObjOffset = 0;
+    CORINFO_OBJECT_HANDLE strObj       = nullptr;
+    if (!GetObjectHandleAndOffset(srcPtr, &strObjOffset, &strObj) || (strObjOffset != OFFSETOF__CORINFO_String__chars))
+    {
+        // TODO: Consider supporting any offset if that is a common pattern, also
+        // static readonly fields (RVA ROS<char> or non-frozen string objects)
+        JITDUMP("GetUtf8Bytes: srcPtr is not a string literal\n")
+        return false;
+    }
+
+    if (info.compCompHnd->getObjectType(strObj) != impGetStringClass())
+    {
+        JITDUMP("GetUtf8Bytes: srcPtr is not a string object\n")
+        return false;
+    }
+
+    if (!srcLen->gtVNPair.BothEqual() || !vnStore->IsVNInt32Constant(srcLen->gtVNPair.GetLiberal()))
+    {
+        JITDUMP("GetUtf8Bytes: srcLen is not constant\n")
+        return false;
+    }
+
+    // Do we need to care if srcLenCns is larger than the actual string literal length? It's a faulty case anyway.
+
+    const int      MaxPossibleUnrollThreshold = 256;
+    const unsigned unrollThreshold            = min(getUnrollThreshold(UnrollKind::Memcpy), MaxPossibleUnrollThreshold);
+    const unsigned srcLenCns                  = (unsigned)vnStore->GetConstantInt32(srcLen->gtVNPair.GetLiberal());
+    if (srcLenCns == 0 || srcLenCns > unrollThreshold)
+    {
+        // TODO: handle srcLenCns == 0 if it's a common case
+        JITDUMP("GetUtf8Bytes: srcLenCns is out of unrollable range\n")
+        return false;
+    }
+
+    // We don't need the length condition if we know that the destination buffer is large enough
+    bool noSizeCheck = false;
+    if (dstLen->gtVNPair.BothEqual() && vnStore->IsVNInt32Constant(dstLen->gtVNPair.GetLiberal()))
+    {
+        const int dstLenCns = vnStore->GetConstantInt32(dstLen->gtVNPair.GetLiberal());
+        noSizeCheck         = dstLenCns >= (int)srcLenCns;
+        assert(dstLenCns > 0);
+    }
+
+    // Read the string literal (UTF16) into a local buffer (UTF8)
+    assert(strObj != nullptr);
+    BYTE srcUtf8cns[MaxPossibleUnrollThreshold] = {}; // same length since we're narrowing U16 to U8
+    for (unsigned charIndex = 0; charIndex < srcLenCns; charIndex++)
+    {
+        uint16_t ch = 0;
+        if (!info.compCompHnd->getStringChar(strObj, (int)charIndex, &ch))
```

Comment:
Seems unfortunate that we have to do a JIT-EE call for every single character. Didn't you add more general purpose API that could be used to bulk copy this data?

Reply:
In SPMI every single of these queries is going to cost us 24+ bytes in storage.

Reply:
Yes, we have `getObjectContent` now